src/target/riscv/batch.c

```diff
@@ -52,7 +55,45 @@ int riscv_batch_run(struct riscv_batch *batch)
 	riscv_batch_add_nop(batch);
 
 	for (size_t i = 0; i < batch->used_scans; ++i) {
-		jtag_add_dr_scan(batch->target->tap, 1, batch->fields + i, TAP_IDLE);
+		if (bscan_tunnel_ir_width != 0) {
```

Comment:
This block looks nearly identical to what we have in dmi_scan(). Can you refactor it to share that code?
(Sorry if I'm repeating myself here. I thought I had commented this earlier but I can't see it anywhere.)

Reply:
Latest commit on this branch factors out the duplicated code, and also factors out the dependent data used by the code (it groups the dependent data into an opaque struct whose instances I figured can live on the stack because jtag_execute_queue is called within the lifetime of the corresponding stack-residing struct in the places it is instantiated/used... the data just needs to be live at least until the associated call to jtag_execute_queue finishes).

Reply:
Actually I think I need to revisit this (in basic testing it seemed to work but I can't explain why), in the batch case I think there needs to be an array of riscv_bscan_tunneled_scan_context_t instances allocated within the batch structure, not just a single instance within riscv_batch_run.  I'll take another look and will probably post another commit.

Reply:
OK I think I got the refactoring correct, it seems to work in focused smoke testing (attach w/ GDB, write some bytes to the load area in flash to clobber some values from any successful previous loads, load a program to same flash area, run compare-sections to verify the load result), but I haven't done a riscv-tests run yet.